src/Storages/StorageS3.cpp

```diff
@@ -182,24 +182,21 @@ String StorageS3Source::DisclosedGlobIterator::next()
 class StorageS3Source::KeysIterator::Impl
 {
 public:
-    explicit Impl(const std::vector<String> & keys_) : keys(keys_), keys_iter(keys.begin())
+    explicit Impl(const std::vector<String> & keys_) : keys(keys_)
     {
     }
 
     String next()
     {
-        std::lock_guard lock(mutex);
-        if (keys_iter == keys.end())
+        size_t current_index = index.fetch_add(1);
```

Comment:
I think that this could be `relaxed` as the threads only modify `index` and only here.